cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -2815,45 +2814,71 @@ const char* soundex_utf8(gdv_int64 ctx, const char* in, gdv_int32 in_len,
   }
 
   // The soundex code is composed by one letter and three numbers
+  char* soundex = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(ctx, in_len));
   char* ret = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(ctx, 4));
-  if (ret == nullptr) {
+
+  if (soundex == nullptr || ret == nullptr) {
     gdv_fn_context_set_error_msg(ctx, "Could not allocate memory for output string");
     *out_len = 0;
     return "";
   }
 
   int si = 1;
+  int ret_len = 1;
   unsigned char c;
 
   int start_idx = 0;
   for (int i = 0; i < in_len; ++i) {
     if (isalpha(in[i]) > 0) {
+      // Retain the first letter
       ret[0] = toupper(in[i]);
       start_idx = i + 1;
       break;
     }
   }
 
-  for (int i = start_idx, l = in_len; i < l; i++) {
+  // If ret[0] is not initialised, return one exception
+  if (start_idx == 0) {
+    gdv_fn_context_set_error_msg(ctx, "There are no valid values in this entry.");
+    *out_len = 0;
+    return "";
+  }
+
+  soundex[0] = '\0';
+  // Replace consonants with digits and special letters with 0
+  for (int i = start_idx; i < in_len; i++) {
     if (isalpha(in[i]) > 0) {
       c = toupper(in[i]) - 65;
-      if (mappings[c] != '0') {
-        if (mappings[c] != ret[si - 1]) {
-          ret[si] = mappings[c];
-          si++;
-        }
-
-        if (si > 3) break;
+      if (mappings[c] != soundex[si - 1]) {
+        soundex[si] = mappings[c];
+        si++;
       }
     }
   }
 
-  if (si <= 3) {
-    while (si <= 3) {
-      ret[si] = '0';
-      si++;
+  int i = 1;
+  // If the saved letter's digit is the same as the resulting first digit, skip it
+  if (soundex[1] == mappings[ret[0] - 65]) {
```

Comment:
What if si = 1? soundex[1] will access an uninitialised variable - you need to handle this and append the remaining 0s and return

Reply:
Yeah, maybe this can broke the code, I added one new verify